In [1]:
import numpy as np
import pandas as pd

In [2]:
data = pd.read_csv('telecom_churn.csv')

In [3]:
data.head()

State  Account length  Area code International plan Voice mail plan  \
0    KS             128        415                 No             Yes   
1    OH             107        415                 No             Yes   
2    NJ             137        415                 No              No   
3    OH              84        408                Yes              No   
4    OK              75        415                Yes              No   

   Number vmail messages  Total day minutes  Total day calls  \
0                     25              265.1              110   
1                     26              161.6              123   
2                      0              243.4              114   
3                      0              299.4               71   
4                      0              166.7              113   

   Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
0             45.07              197.4               99             16.78   
1             27.47              195.5              103             16.62   
2             41.38              121.2              110             10.30   
3             50.90               61.9               88              5.26   
4             28.34              148.3              122             12.61   

   Total night minutes  Total night calls  Total night charge  \
0                244.7                 91               11.01   
1                254.4                103               11.45   
2                162.6                104                7.32   
3                196.9                 89                8.86   
4                186.9                121                8.41   

   Total intl minutes  Total intl calls  Total intl charge  \
0                10.0                 3               2.70   
1                13.7                 3               3.70   
2                12.2                 5               3.29   
3                 6.6                 7               1.78   
4                10.1                 3               2.73   

   Customer service calls  Churn  
0                       1  False  
1                       1  False  
2                       0  False  
3                       2  False  
4                       3  False

In [4]:
print(data.shape, '\n')
print(data.columns, '\n')
print(data.info(), '\n')

(3333, 20) 

Index(['State', 'Account length', 'Area code', 'International plan',
       'Voice mail plan', 'Number vmail messages', 'Total day minutes',
       'Total day calls', 'Total day charge', 'Total eve minutes',
       'Total eve calls', 'Total eve charge', 'Total night minutes',
       'Total night calls', 'Total night charge', 'Total intl minutes',
       'Total intl calls', 'Total intl charge', 'Customer service calls',
       'Churn'],
      dtype='object') 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3333 entries, 0 to 3332
Data columns (total 20 columns):
State                     3333 non-null object
Account length            3333 non-null int64
Area code                 3333 non-null int64
International plan        3333 non-null object
Voice mail plan           3333 non-null object
Number vmail messages     3333 non-null int64
Total day minutes         3333 non-null float64
Total day calls           3333 non-null int64
Total day charge          3333 non-null flo

In [5]:
data.describe()

Account length    Area code  Number vmail messages  Total day minutes  \
count     3333.000000  3333.000000            3333.000000        3333.000000   
mean       101.064806   437.182418               8.099010         179.775098   
std         39.822106    42.371290              13.688365          54.467389   
min          1.000000   408.000000               0.000000           0.000000   
25%         74.000000   408.000000               0.000000         143.700000   
50%        101.000000   415.000000               0.000000         179.400000   
75%        127.000000   510.000000              20.000000         216.400000   
max        243.000000   510.000000              51.000000         350.800000   

       Total day calls  Total day charge  Total eve minutes  Total eve calls  \
count      3333.000000       3333.000000        3333.000000      3333.000000   
mean        100.435644         30.562307         200.980348       100.114311   
std          20.069084          9.259435          50.713844        19.922625   
min           0.000000          0.000000           0.000000         0.000000   
25%          87.000000         24.430000         166.600000        87.000000   
50%         101.000000         30.500000         201.400000       100.000000   
75%         114.000000         36.790000         235.300000       114.000000   
max         165.000000         59.640000         363.700000       170.000000   

       Total eve charge  Total night minutes  Total night calls  \
count       3333.000000          3333.000000        3333.000000   
mean          17.083540           200.872037         100.107711   
std            4.310668            50.573847          19.568609   
min            0.000000            23.200000          33.000000   
25%           14.160000           167.000000          87.000000   
50%           17.120000           201.200000         100.000000   
75%           20.000000           235.300000         113.000000   
max           30.910000           395.000000         175.000000   

       Total night charge  Total intl minutes  Total intl calls  \
count         3333.000000         3333.000000       3333.000000   
mean             9.039325           10.237294          4.479448   
std              2.275873            2.791840          2.461214   
min              1.040000            0.000000          0.000000   
25%              7.520000            8.500000          3.000000   
50%              9.050000           10.300000          4.000000   
75%             10.590000           12.100000          6.000000   
max             17.770000           20.000000         20.000000   

       Total intl charge  Customer service calls  
count        3333.000000             3333.000000  
mean            2.764581                1.562856  
std             0.753773                1.315491  
min             0.000000                0.000000  
25%             2.300000                1.000000  
50%             2.780000                1.000000  
75%             3.270000                2.000000  
max             5.400000                9.000000

Чтобы посмотреть статистику по нечисловым признакам, нужно явно указать интересующие нас типы в параметре include:

In [6]:
data.describe(include=['object', 'bool'])

State International plan Voice mail plan  Churn
count   3333               3333            3333   3333
unique    51                  2               2      2
top       WV                 No              No  False
freq     106               3010            2411   2850

Для категориальных (тип object) и булевых (тип bool) признаков можно воспользоваться методом value_counts. Посмотрим на распределение данных по нашей целевой переменной — Churn:

In [7]:
data['Churn'].value_counts()

False    2850
True      483
Name: Churn, dtype: int64

Укажем значение параметра normalize=True, чтобы посмотреть не абсолютные частоты, а относительные:

In [8]:
data['Area code'].value_counts(normalize=True)

415    0.496550
510    0.252025
408    0.251425
Name: Area code, dtype: float64

In [9]:
data['Area code'].value_counts()

415    1655
510     840
408     838
Name: Area code, dtype: int64

DataFrame можно отсортировать по значению какого-нибудь из признаков:

In [10]:
data.sort_values(by='Total day charge', 
                ascending=False).head(10)

State  Account length  Area code International plan Voice mail plan  \
365     CO             154        415                 No              No   
985     NY              64        415                Yes              No   
2594    OH             115        510                Yes              No   
156     OH              83        415                 No              No   
605     MO             112        415                 No              No   
975     DE             129        510                 No              No   
15      NY             161        415                 No              No   
1679    OK              82        510                 No              No   
2028    SD              93        510                 No              No   
197     TX             208        510                 No              No   

      Number vmail messages  Total day minutes  Total day calls  \
365                       0              350.8               75   
985                       0              346.8               55   
2594                      0              345.3               81   
156                       0              337.4              120   
605                       0              335.5               77   
975                       0              334.3              118   
15                        0              332.9               67   
1679                      0              329.8               73   
2028                      0              328.1              106   
197                       0              326.5               67   

      Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
365              59.64              216.5               94             18.40   
985              58.96              249.5               79             21.21   
2594             58.70              203.4              106             17.29   
156              57.36              227.4              116             19.33   
605              57.04              212.5              109             18.06   
975              56.83              192.1              104             16.33   
15               56.59              317.8               97             27.01   
1679             56.07              208.3              120             17.71   
2028             55.78              151.7               89             12.89   
197              55.51              176.3              113             14.99   

      Total night minutes  Total night calls  Total night charge  \
365                 253.9                100               11.43   
985                 275.4                102               12.39   
2594                217.5                107                9.79   
156                 153.9                114                6.93   
605                 265.0                132               11.93   
975                 191.0                 83                8.59   
15                  160.6                128                7.23   
1679                267.1                102               12.02   
2028                303.5                114               13.66   
197                 181.7                102                8.18   

      Total intl minutes  Total intl calls  Total intl charge  \
365                 10.1                 9               2.73   
985                 13.3                 9               3.59   
2594                11.8                 8               3.19   
156                 15.8                 7               4.27   
605                 12.7                 8               3.43   
975                 10.4                 6               2.81   
15                   5.4                 9               1.46   
1679                10.6                 6               2.86   
2028                 8.7                 3               2.35   
197                 10.7                 6               2.89   

      Customer service calls  Churn  
365                        1   True  
985     

Сортировать можно и по группе столбцов:

In [11]:
data.sort_values(by=['Churn', 'Total day charge'],
                ascending=[True, False]).head(10)

State  Account length  Area code International plan Voice mail plan  \
688     MN              13        510                 No             Yes   
2259    NC             210        415                 No             Yes   
534     LA              67        510                 No              No   
575     SD             114        415                 No             Yes   
2858    AL             141        510                 No             Yes   
1165    VT              50        415                Yes             Yes   
3207    DC              93        408                 No             Yes   
2478    TN             123        415                 No             Yes   
1355    TX              50        408                 No             Yes   
1830    CT              50        408                 No              No   

      Number vmail messages  Total day minutes  Total day calls  \
688                      21              315.6              105   
2259                     31              313.8               87   
534                       0              310.4               97   
575                      36              309.9               90   
2858                     28              308.0              123   
1165                     26              307.1               94   
3207                     22              306.2              123   
2478                     34              305.2               80   
1355                     31              302.7               93   
1830                      0              301.7               82   

      Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
688              53.65              208.9               71             17.76   
2259             53.35              147.7              103             12.55   
534              52.77               66.5              123              5.65   
575              52.68              200.3               89             17.03   
2858             52.36              247.8              128             21.06   
1165             52.21              289.4               78             24.60   
3207             52.05              189.7               83             16.12   
2478             51.88              156.5              109             13.30   
1355             51.46              240.5              119             20.44   
1830             51.29              167.1              118             14.20   

      Total night minutes  Total night calls  Total night charge  \
688                 260.1                123               11.70   
2259                192.7                 97                8.67   
534                 246.5                 99               11.09   
575                 183.5                105                8.26   
2858                152.9                103                6.88   
1165                174.9                109                7.87   
3207                240.3                107               10.81   
2478                280.0                 81               12.60   
1355                193.9                103                8.73   
1830                 72.2                 89                3.25   

      Total intl minutes  Total intl calls  Total intl charge  \
688                 12.1                 3               3.27   
2259                10.1                 7               2.73   
534                  9.2                10               2.48   
575                 14.2                 2               3.83   
2858                 7.4                 3               2.00   
1165                 8.0                 3               2.16   
3207                11.7                 2               3.16   
2478                13.2                 7               3.56   
1355                13.6                14               3.67   
1830                10.5                 6               2.84   

      Customer service calls  Churn  
688                        3  False  
2259    

----------
-----------
-------

Какова доля людей нелояльных пользователей в нашем датафрейме?

In [12]:
data['Churn'].mean()

0.14491449144914492

Средние значения числовых признаков среди нелояльных пользователей:

In [13]:
data[data['Churn'] == 1].mean()

Account length            102.664596
Area code                 437.817805
Number vmail messages       5.115942
Total day minutes         206.914079
Total day calls           101.335404
Total day charge           35.175921
Total eve minutes         212.410145
Total eve calls           100.561077
Total eve charge           18.054969
Total night minutes       205.231677
Total night calls         100.399586
Total night charge          9.235528
Total intl minutes         10.700000
Total intl calls            4.163561
Total intl charge           2.889545
Customer service calls      2.229814
Churn                       1.000000
dtype: float64

Сколько в среднем в течение дня разговаривают по телефону нелояльные пользователи?

In [14]:
data[data['Churn'] == 1]['Total day minutes'].mean()

206.91407867494814

Датафреймы можно индексировать как по названию столбца или строки, так и по порядковому номеру. <br>Для индексации по названию используется метод loc, по номеру — iloc.

В первом случае мы говорим «передай нам значения для id строк от 0 до 5 и для столбцов от State до Area code», <br>а во втором — «передай нам значения первых пяти строк в первых трёх столбцах».

Когда мы передаём slice object в iloc, датафрейм слайсится как обычно. <br>Однако в случае с loc учитываются и начало, и конец слайса.

In [15]:
data.loc[1:10, 'Area code':'Number vmail messages']

Area code International plan Voice mail plan  Number vmail messages
1         415                 No             Yes                     26
2         415                 No              No                      0
3         408                Yes              No                      0
4         415                Yes              No                      0
5         510                Yes              No                      0
6         510                 No             Yes                     24
7         415                Yes              No                      0
8         408                 No              No                      0
9         415                Yes             Yes                     37
10        415                 No              No                      0

In [16]:
data.iloc[1:10, 2:5]

Area code International plan Voice mail plan
1        415                 No             Yes
2        415                 No              No
3        408                Yes              No
4        415                Yes              No
5        510                Yes              No
6        510                 No             Yes
7        415                Yes              No
8        408                 No              No
9        415                Yes             Yes

Если нам нужна первая или последняя строчка датафрейма:

In [17]:
data[-1:]

State  Account length  Area code International plan Voice mail plan  \
3332    TN              74        415                 No             Yes   

      Number vmail messages  Total day minutes  Total day calls  \
3332                     25              234.4              113   

      Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
3332             39.85              265.9               82              22.6   

      Total night minutes  Total night calls  Total night charge  \
3332                241.4                 77               10.86   

      Total intl minutes  Total intl calls  Total intl charge  \
3332                13.7                 4                3.7   

      Customer service calls  Churn  
3332                       0  False

-----------
-----------
-----------

Применение функции к каждому столбцу: apply.

Метод apply можно использовать и для того, чтобы применить функцию к каждой строке. <br>Для этого нужно указать axis=1.

In [18]:
data.apply(np.max) 

State                        WY
Account length              243
Area code                   510
International plan          Yes
Voice mail plan             Yes
Number vmail messages        51
Total day minutes         350.8
Total day calls             165
Total day charge          59.64
Total eve minutes         363.7
Total eve calls             170
Total eve charge          30.91
Total night minutes         395
Total night calls           175
Total night charge        17.77
Total intl minutes           20
Total intl calls             20
Total intl charge           5.4
Customer service calls        9
Churn                      True
dtype: object

Применение функции к каждой ячейке столбца: map.

Например, метод map можно использовать для замены значений в колонке, <br>передав ему в качестве аргумента словарь вида {old_value: new_value}:

In [19]:
d = {'No' : False, 'Yes' : True}

data['International plan'] = data['International plan'].map(d)
data.head(10)

State  Account length  Area code  International plan Voice mail plan  \
0    KS             128        415               False             Yes   
1    OH             107        415               False             Yes   
2    NJ             137        415               False              No   
3    OH              84        408                True              No   
4    OK              75        415                True              No   
5    AL             118        510                True              No   
6    MA             121        510               False             Yes   
7    MO             147        415                True              No   
8    LA             117        408               False              No   
9    WV             141        415                True             Yes   

   Number vmail messages  Total day minutes  Total day calls  \
0                     25              265.1              110   
1                     26              161.6              123   
2                      0              243.4              114   
3                      0              299.4               71   
4                      0              166.7              113   
5                      0              223.4               98   
6                     24              218.2               88   
7                      0              157.0               79   
8                      0              184.5               97   
9                     37              258.6               84   

   Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
0             45.07              197.4               99             16.78   
1             27.47              195.5              103             16.62   
2             41.38              121.2              110             10.30   
3             50.90               61.9               88              5.26   
4             28.34              148.3              122             12.61   
5             37.98              220.6              101             18.75   
6             37.09              348.5              108             29.62   
7             26.69              103.1               94              8.76   
8             31.37              351.6               80             29.89   
9             43.96              222.0              111             18.87   

   Total night minutes  Total night calls  Total night charge  \
0                244.7                 91               11.01   
1                254.4                103               11.45   
2                162.6                104                7.32   
3                196.9                 89                8.86   
4                186.9                121                8.41   
5                203.9                118                9.18   
6                212.6                118                9.57   
7                211.8                 96                9.53   
8                215.8                 90                9.71   
9                326.4                 97               14.69   

   Total intl minutes  Total intl calls  Total intl charge  \
0                10.0                 3               2.70   
1                13.7                 3               3.70   
2                12.2                 5               3.29   
3                 6.6                 7               1.78   
4                10.1                 3               2.73   
5                 6.3                 6               1.70   
6                 7.5                 7               2.03   
7                 7.1                 6               1.92   
8                 8.7                 4               2.35   
9                11.2                 5               3.02   

   Customer service calls  Churn  
0                       1  False  
1                       1  False  
2                       0  False  
3                       2  False  
4                       3  False  
5                       0

Аналогичную процедуру можно провернуть с помощью метода replace:

In [20]:
data = data.replace({'Voice mail plan': d})
data.head(10)

State  Account length  Area code  International plan  Voice mail plan  \
0    KS             128        415               False             True   
1    OH             107        415               False             True   
2    NJ             137        415               False            False   
3    OH              84        408                True            False   
4    OK              75        415                True            False   
5    AL             118        510                True            False   
6    MA             121        510               False             True   
7    MO             147        415                True            False   
8    LA             117        408               False            False   
9    WV             141        415                True             True   

   Number vmail messages  Total day minutes  Total day calls  \
0                     25              265.1              110   
1                     26              161.6              123   
2                      0              243.4              114   
3                      0              299.4               71   
4                      0              166.7              113   
5                      0              223.4               98   
6                     24              218.2               88   
7                      0              157.0               79   
8                      0              184.5               97   
9                     37              258.6               84   

   Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
0             45.07              197.4               99             16.78   
1             27.47              195.5              103             16.62   
2             41.38              121.2              110             10.30   
3             50.90               61.9               88              5.26   
4             28.34              148.3              122             12.61   
5             37.98              220.6              101             18.75   
6             37.09              348.5              108             29.62   
7             26.69              103.1               94              8.76   
8             31.37              351.6               80             29.89   
9             43.96              222.0              111             18.87   

   Total night minutes  Total night calls  Total night charge  \
0                244.7                 91               11.01   
1                254.4                103               11.45   
2                162.6                104                7.32   
3                196.9                 89                8.86   
4                186.9                121                8.41   
5                203.9                118                9.18   
6                212.6                118                9.57   
7                211.8                 96                9.53   
8                215.8                 90                9.71   
9                326.4                 97               14.69   

   Total intl minutes  Total intl calls  Total intl charge  \
0                10.0                 3               2.70   
1                13.7                 3               3.70   
2                12.2                 5               3.29   
3                 6.6                 7               1.78   
4                10.1                 3               2.73   
5                 6.3                 6               1.70   
6                 7.5                 7               2.03   
7                 7.1                 6               1.92   
8                 8.7                 4               2.35   
9                11.2                 5               3.02   

   Customer service calls  Churn  
0                       1  False  
1                       1  False  
2                       0  False  
3                       2  False  
4                       3  False  
5             

Группирование данных в зависимости от значения признака Churn и вывод статистик по трём столбцам в каждой группе:

In [21]:
columns_to_show = ['Total day minutes', 'Total eve minutes', 'Total night minutes']

data.groupby(['Churn'])[columns_to_show].describe(percentiles=[])

Total day minutes                                            \
                  count        mean        std  min    50%    max   
Churn                                                               
False            2850.0  175.175754  50.181655  0.0  177.2  315.6   
True              483.0  206.914079  68.997792  0.0  217.6  350.8   

      Total eve minutes                                             \
                  count        mean        std   min    50%    max   
Churn                                                                
False            2850.0  199.043298  50.292175   0.0  199.6  361.8   
True              483.0  212.410145  51.728910  70.9  211.3  363.7   

      Total night minutes                                              
                    count        mean        std   min     50%    max  
Churn                                                                  
False              2850.0  200.133193  51.105032  23.2  200.25  395.0  
True                483.0  205.231677  47.132825  47.4  204.80  354.9

Сделаем то же самое, но немного по-другому, передав в agg список функций:

In [22]:
data.groupby(['Churn'])[columns_to_show].agg([np.mean, np.std, np.min, np.max])

Total day minutes                        Total eve minutes             \
                   mean        std amin   amax              mean        std   
Churn                                                                         
False        175.175754  50.181655  0.0  315.6        199.043298  50.292175   
True         206.914079  68.997792  0.0  350.8        212.410145  51.728910   

                   Total night minutes                          
       amin   amax                mean        std  amin   amax  
Churn                                                           
False   0.0  361.8          200.133193  51.105032  23.2  395.0  
True   70.9  363.7          205.231677  47.132825  47.4  354.9

--------------
-------
-----

Допустим, мы хотим посмотреть, как наблюдения в нашей выборке распределены в контексте двух признаков — Churn и International plan. <br>Для этого мы можем построить таблицу сопряженности, воспользовавшись методом crosstab:

In [23]:
pd.crosstab(data['Churn'], data['International plan'])

International plan  False  True 
Churn                           
False                2664    186
True                  346    137

Таблица сопряжённости — средство представления совместного распределения двух величин, <br>предназначенное для исследования связи между ними.

На пересечении строки и столбца указывается частота совместного появления $f_{ij}$ соответствующих значений двух признаков $x_{i}$ и $y_{j}$. Сумма частот по строке $f_{i}$. называется маргинальной частотой строки; сумма частот по столбцу $f_{.j}$ — маргинальной частотой столбца. Сумма маргинальных частот равна объёму выборки $n$; их распределение представляет собой одномерное распределение переменной, образующей строки или столбцы таблицы.

In [24]:
pd.crosstab(data['Churn'], data['International plan'], normalize=True)

International plan    False     True 
Churn                                
False               0.79928  0.055806
True                0.10381  0.041104

Сводные таблицы помогают резюмировать большие объёмы данных в сравнительной таблице.

В Pandas за сводные таблицы отвечает метод pivot_table, который принимает в качестве параметров:

* values – список переменных, по которым требуется рассчитать нужные статистики,
* index – список переменных, по которым нужно сгруппировать данные,
* aggfunc — то, что нам, собственно, нужно посчитать по группам — сумму, среднее, максимум, минимум или что-то ещё.

Давайте посмотрим среднее число дневных, вечерних и ночных звонков для разных Area code:

In [25]:
data.pivot_table(['Total day calls', 'Total eve calls', 'Total night calls'], 
                 ['Area code'], 
                 aggfunc='mean').head(10)

Total day calls  Total eve calls  Total night calls
Area code                                                     
408             100.496420        99.788783          99.039379
415             100.576435       100.503927         100.398187
510             100.097619        99.671429         100.601190

-----
-------
------

Посмотрим, как отток связан с признаком "Подключение международного роуминга" (International plan). <br>Сделаем это с помощью таблички crosstab:

In [26]:
pd.crosstab(data['Churn'], data['International plan'], margins=True)

International plan  False  True   All
Churn                                
False                2664   186  2850
True                  346   137   483
All                  3010   323  3333

Видим, что когда роуминг подключен, доля оттока намного выше – интересное наблюдение!

Посмотрим на еще один важный признак – "Число обращений в сервисный центр" (Customer service calls):

In [27]:
pd.crosstab(data['Churn'], data['Customer service calls'], margins=True)

Customer service calls    0     1    2    3    4   5   6  7  8  9   All
Churn                                                                  
False                   605  1059  672  385   90  26   8  4  1  0  2850
True                     92   122   87   44   76  40  14  5  1  2   483
All                     697  1181  759  429  166  66  22  9  2  2  3333

Доля оттока значительно возрастает начиная с 4 звонков в сервисный центр.

Добавим теперь в наш DataFrame бинарный признак — результат сравнения Customer service calls > 3. <br>И еще раз посмотрим, как он связан с оттоком:

In [28]:
data['Many_service_calls'] = (data['Customer service calls'] > 3).astype('bool')

pd.crosstab(data['Many_service_calls'], data['Churn'], margins=True)

Churn               False  True   All
Many_service_calls                   
False                2721   345  3066
True                  129   138   267
All                  2850   483  3333

Объединим рассмотренные выше условия и построим сводную табличку для этого объединения и оттока:

In [29]:
pd.crosstab(data['Many_service_calls'] & data['International plan'] , data['Churn'])

Churn  False  True 
row_0              
False   2841    464
True       9     19

Значит, прогнозируя отток клиента в случае, когда число звонков в сервисный центр больше 3 и подключен роуминг <br>(и прогнозируя лояльность – в противном случае), можно ожидать около 85.8% правильных попаданий (ошибаемся всего 464 + 9 раз).

Доля лояльных клиентов в выборке – 85.5%. Самая наивная модель, ответ которой "клиент всегда лоялен" <br>на подобных данных будет угадывать примерно в 85.5% случаев. То есть доли правильных ответов (accuracy) <br>последующих моделей должны быть как минимум не меньше, а лучше, значительно выше этой цифры.